cpp/oneapi/dal/algo/decision_forest/backend/gpu/train_best_split_sp_opt_impl_dpc.cpp

```diff
@@ -0,0 +1,1143 @@
+                        &local_byte_buf_ptr,
+                        target_sbg_count * split_info_t::get_cache_without_hist_byte_size());
+
+                    if (sub_group_local_id == 0) {
+                        bs.store_without_hist(local_buf_ptr, sub_group_id, target_sbg_count);
+                    }
+                    item.barrier(sycl::access::fence_space::local_space);
+
+                    if (sub_group_id == 0) {
+                        bs.load_without_hist(local_buf_ptr,
+                                             sub_group_local_id % target_sbg_count,
+                                             target_sbg_count);
+                        bs.set_left_hist_ptr(local_bs_hist_buf_ptr +
+                                             (sub_group_local_id % target_sbg_count) *
+                                                 bs.hist_prop_count);
+
+                        if (sp_hlp
+                                .my_split_is_best_for_sbg(item, bs, node_ptr, node_id, index_max)) {
+                            sp_hlp.update_node_bs_info(bs,
+                                                       node_ptr,
+                                                       node_imp_decr_list_ptr,
+                                                       node_id,
+                                                       index_max,
+                                                       update_imp_dec_required);
+                            if constexpr (std::is_same_v<Task, task::classification>) {
+                                sp_hlp.update_left_child_imp(left_child_imp_list_ptr,
+                                                             left_child_class_hist_list_ptr,
+                                                             bs.left_imp,
+                                                             bs.left_hist,
+                                                             node_id,
+                                                             class_count);
+                            }
+                            else {
+                                sp_hlp.update_left_child_imp(left_child_imp_list_ptr,
+                                                             bs.left_hist,
+                                                             node_id);
+                            }
+                        }
+                    }
+                });
+        });
+
+        last_event.wait_and_throw();
+    }
+
+    return last_event;
+}
+
+#define INSTANTIATE(F, B, I, T) template class train_best_split_sp_opt_impl<F, B, I, T>;
+
+INSTANTIATE(float, std::uint32_t, std::int32_t, task::classification);
+INSTANTIATE(float, std::uint32_t, std::int32_t, task::regression);
+
+INSTANTIATE(double, std::uint32_t, std::int32_t, task::classification);
+INSTANTIATE(double, std::uint32_t, std::int32_t, task::regression);
+
+} // namespace oneapi::dal::decision_forest::backend
+
+#endif
```

Comment:
This file is too big for me

Reply:
moved all helpers into separate train_best_split_helpers.hpp file